### PR TITLE
A new approach to Stepper and Stream materialization

### PR DESCRIPTION
**Pull request by szeiger**
Getting the most efficient Stepper implementation requires knowledge
of the static type of a collection. This problem could be solved in the
most elegant way by integrating Steppers into the collections framework
as a replacement for Iterators (i.e. every collection gets a `stepper`
method and `iterator` delegates to `stepper` by default).

But this is at odds with the handling of specialized primitive Steppers
in the current implementation. If we want `stepper` to be an instance
methods instead of an extension method, there needs to be a single such
method for all specialized Steppers. The fundamental change in this
new implementation is to encode the translation from element type to
Stepper type (including widening conversions) as a functional dependency
via the new `StepperShape` trait.

This greatly reduces the number of implicit methods and classes and
keeps all specialized versions of `MakesStepper` together. The default
base classes support all unboxing and widening conversions so that a
simple `MakesStepper` for a collection of boxed elements only needs to
handle the `AnyStepper` case.

`keyStepper` and `valueStepper` are handled in the same way.

`StreamConverters` use a separate `StreamShape` for the translation from
element type to `BaseStream` subtype which is compatible with
`StepConverters` and supports the same primitive types and widening
conversions.

### CODE REVIEW

**Comment by szeiger** [comment]
@Ichoran, do you have time to review this?

**Comment by Ichoran** [comment]
I can on Wednesday morning.

File: src/main/scala/scala/compat/java8/StreamConverters.scala [hunk 1]
```diff
@@ -6,182 +6,85 @@ import java.util.stream._
...
-  }
-  implicit class EnrichBoxedShortKeySteppableWithSeqKeyStream[CC](cc: CC)(implicit steppize: CC => MakesKeyStepper[AnyStepper[Short]]) {
-    def seqKeyStream: IntStream = StreamSupport.intStream(new Stepper.WideningShortStepper(steppize(cc).keyStepper), false)
+  // widening
+  implicit val ByteValue    = intStreamShape[Byte]
+  implicit val ShortValue   = intStreamShape[Short]
+  implicit val CharValue    = intStreamShape[Char]
+  implicit val FloatValue   = doubleStreamShape[Float]
+}
+trait StreamShapeLowPrio {
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/StreamConverters.scala]: Can we spell out "Priority" here?

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 2]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
+  implicit val LongValue    = valueShape[Long, LongStepper]
+  implicit val DoubleValue  = valueShape[Double, DoubleStepper]
+
+  // widening
+  implicit val ByteValue    = valueShape[Byte, IntStepper]
+  implicit val ShortValue   = valueShape[Short, IntStepper]
+  implicit val CharValue    = valueShape[Char, IntStepper]
+  implicit val FloatValue   = valueShape[Float, DoubleStepper]
+}
+trait StepperShapeLowPrio {
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: Again, don't think it's worth the potential loss of clarity to save the characters `rity`.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 3]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
-  def parStream: SS
+trait MakesKeyValueStepper[K, V, +Extra] extends Any {
+  /** Generates a fresh stepper of type `S` over map keys of type `K` */
+  def keyStepper[S <: Stepper[_]](implicit ss: StepperShape[K, S]): S with Extra
+
+  /** Generates a fresh stepper of type `S` over map values of type `V` */
+  def valueStepper[S <: Stepper[_]](implicit ss: StepperShape[V, S]): S with Extra
+}
+
+sealed trait StepperShape[T, S <: Stepper[_]] { def ref: Boolean }
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: Can you document what `ref` does, since it's public?

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 4]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
 }
 
-trait MakesStepper[+T <: Stepper[_]] extends Any {
-  /** Generates a fresh stepper of type `T` */
-  def stepper: T
+/** Superclass for `MakesStepper` implementations which support parallelization. At least the `AnyStepper` case must be
+  * implemented, all others default to building an `AnyStepper` and putting an unboxing conversion on top. */
+trait MakesParStepper[T] extends Any with MakesStepper[T, EfficientSubstep] {
+  def stepper[S <: Stepper[_]](implicit ss: StepperShape[T, S]) = (ss match {
+    case StepperShape.IntValue    => new Stepper.UnboxingIntStepper   (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Int   ]]) with EfficientSubstep
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: This is a pretty big case statement.  Can you add a method to StepperShape that evokes the right boxing?  I'm pretty sure that multiple dispatch is going to be faster in this case, and it'd certainly help with DRY.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 5]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
+    case StepperShape.LongValue   => new Stepper.UnboxingLongStepper  (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Long  ]]) with EfficientSubstep
+    case StepperShape.DoubleValue => new Stepper.UnboxingDoubleStepper(stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Double]]) with EfficientSubstep
+    case StepperShape.ByteValue   => new Stepper.UnboxingByteStepper  (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Byte  ]]) with EfficientSubstep
+    case StepperShape.ShortValue  => new Stepper.UnboxingShortStepper (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Short ]]) with EfficientSubstep
+    case StepperShape.CharValue   => new Stepper.UnboxingCharStepper  (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Char  ]]) with EfficientSubstep
+    case StepperShape.FloatValue  => new Stepper.UnboxingFloatStepper (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Float ]]) with EfficientSubstep
+    case _                        => throw new NotImplementedError("AnyStepper must be handled in `stepper` implementations")
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: This seems too fragile for something that users can extend.  Can you maybe have an extra abstract `anyStepper` method that this defers to (so the user must fill it in), and then at least users won't accidentally fail to realize that they need to do anything (or some particular thing) and then have it blow up at runtime?

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 6]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
+  }).asInstanceOf[S with EfficientSubstep]
 }
 
-trait MakesKeyStepper[+T <: Stepper[_]] extends Any {
-  /** Generates a fresh stepper of type `T` over map keys */
-  def keyStepper: T
+/** Superclass for `MakesStepper` implementations which do not support parallelization. At least the `AnyStepper` case must be
+  * implemented, all others default to building an `AnyStepper` and putting an unboxing conversion on top. */
+trait MakesSeqStepper[T] extends Any with MakesStepper[T, Any] {
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: Can we DRY this more by adding another trait that has the implementation parameterized on the return type?  AFAICT it is exactly the same code.  (Then again, if `StepperShape` handles it, maybe there won't be anything else to do.)

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala [hunk 7]
```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
...
+    case StepperShape.DoubleValue => new Stepper.UnboxingDoubleStepper(keyStepper(StepperShape.anyStepperShape[K]).asInstanceOf[AnyStepper[Double]]) with EfficientSubstep
+    case StepperShape.ByteValue   => new Stepper.UnboxingByteStepper  (keyStepper(StepperShape.anyStepperShape[K]).asInstanceOf[AnyStepper[Byte  ]]) with EfficientSubstep
+    case StepperShape.ShortValue  => new Stepper.UnboxingShortStepper (keyStepper(StepperShape.anyStepperShape[K]).asInstanceOf[AnyStepper[Short ]]) with EfficientSubstep
+    case StepperShape.CharValue   => new Stepper.UnboxingCharStepper  (keyStepper(StepperShape.anyStepperShape[K]).asInstanceOf[AnyStepper[Char  ]]) with EfficientSubstep
+    case StepperShape.FloatValue  => new Stepper.UnboxingFloatStepper (keyStepper(StepperShape.anyStepperShape[K]).asInstanceOf[AnyStepper[Float ]]) with EfficientSubstep
+    case _                        => throw new NotImplementedError("AnyStepper case must be handled in `keyStepper` implementations")
+  }).asInstanceOf[S with EfficientSubstep]
+
+  def valueStepper[S <: Stepper[_]](implicit ss: StepperShape[V, S]) = (ss match {
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: I think this whole thing could be just `ss.unboxing(valueStepper(StepperShape.anyStepperShape[V]))`

**Comment by Ichoran** [comment]
Partway through the review, but I'm out of time for right now.  I'll try to finish later tonight or, failing that, (my) Wednesday morning.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala (hunk 4 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: This can't be in StepperShape because all implementations are slightly different (calling `stepper`, `keyStepper` or `valueStepper`, creating a wrapper with out without `EfficientSubstep`). We could potentially improve performance by dispatching on an Int value representing the StepperShape instead of using the object identities (like `StepperShape.IntValue`) so that a TABLESWITCH can be used.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala (hunk 5 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: I tried this at first but abandoned it for some reason. I can try again.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala (hunk 6 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: We need to create instances of different Steppers either `with EfficientSubstep` or without. There's no way to abstract over that.

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala (hunk 7 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: Hm, that could actually work (with two different methods for seq/par)

File: src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala (hunk 5 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: I tried again and now I remember why I didn't do it this way. There are types like `BitSet` for which an `AnyStepper` doesn't make sense in the current implementation because the element type is statically known to be `Int`. Consequently there is no `AnyStepper` implementation for `BitSet`. The situation would be different if we put the `stepper` methods directly into collection classes. `BitSet`'s `stepper` methods would have to take the situation into account that a `BitSet` is for example seen as a `Set[_]`, so it would have to return an `AnyStepper[_]`, but that just means that the `stepper` implementation needs to check for the type and call `super.stepper` if it's not recognized. `BitSet` would then inherit the `AnyStepper` implementation from its parent class.

Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala]: On second thought, there's no reason why `RichBitSetCanStep` would want to extend `MakesParStepper[Int]` instead of `MakesStepper[Int, EfficientSubstep]`. Either you still call `super.stepper`, then you need the `AnyStepper` version, or you don't, then you don't have to extend the abstract implementations.

**Comment by szeiger** [comment]
Updated based on review comments.

File: src/main/scala/scala/compat/java8/StreamConverters.scala [hunk 8]
```diff
@@ -6,182 +6,85 @@ import java.util.stream._
...
-    def seqStream: LongStream = StreamSupport.longStream(steppize(cc).stepper, false)
-  }
-  implicit class EnrichScalaCollectionWithSeqDoubleKeyStream[CC](cc: CC)(implicit steppize: CC => MakesKeyStepper[DoubleStepper]) {
-    def seqKeyStream: DoubleStream = StreamSupport.doubleStream(steppize(cc).keyStepper, false)
+trait PrimitiveStreamAccumulator[S, AA] {
+  def streamAccumulate(stream: S): AA
+}
+
+trait PrimitiveStreamUnboxer[A, S] {
+  def apply(boxed: Stream[A]): S
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/StreamConverters.scala]: I'm not sure `apply` is the best choice for a method name, since it makes it impossible (with that signature) to have something that is both a function and a `PrimitiveStreamUnboxer`.  If it's for internal use only, just seal it and it's fine.  Otherwise I would pick something like `streamUnbox` to mirror `streamAccumulate` above.

File: src/main/scala/scala/compat/java8/converterImpl/StepsArray.scala [hunk 9]
```diff
@@ -81,46 +70,18 @@ extends StepsLongLikeIndexed[StepsLongArray](_i0, _iN) {
...
-}
-
-final class RichArrayLongCanStep(private val underlying: Array[Long]) extends AnyVal with MakesStepper[LongStepper with EfficientSubstep] {
-  @inline def stepper: LongStepper with EfficientSubstep = new StepsLongArray(underlying, 0, underlying.length)
+final class RichArrayCanStep[T](private val underlying: Array[T]) extends AnyVal with MakesStepper[T, EfficientSubstep] {
+  override def stepper[S <: Stepper[_]](implicit ss: StepperShape[T, S]) = ((ss.shape: @switch) match {
+    case StepperShape.Reference =>
+      if(underlying.isInstanceOf[Array[Boolean]])
+                                     new StepsBoxedBooleanArray  (underlying.asInstanceOf[Array[Boolean]], 0, underlying.length)
+      else                           new StepsObjectArray[AnyRef](underlying.asInstanceOf[Array[AnyRef ]], 0, underlying.length)
```
Comment (Ichoran) [inline src/main/scala/scala/compat/java8/converterImpl/StepsArray.scala]: I know it's not very useful to stream Unit, but will the `AnyRef` case catch it fine?  It's certainly a problem if you type something as `AnyRef` when it's actually `[Lscala.runtime.BoxedUnit`, but I think we will be read-only here so it's okay?  You at least should leave a comment explaining why you think it's okay.

**Comment by Ichoran** [comment]
Aside for the few extra comments I left, this looks great!  Definitely a more compact design.

Have you run the benchmarks again to see if performance stays comparable (especially for small collections)?  There's always a concern that even though the extra abstraction is logically entirely removable that in practice it will cause a performance hit.

File: src/main/scala/scala/compat/java8/StreamConverters.scala (hunk 8 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/StreamConverters.scala]: This is from your original PR, it just got moved around. I haven't really looked into this area of the codebase yet after the initial review.

File: src/main/scala/scala/compat/java8/converterImpl/StepsArray.scala (hunk 9 above)
Comment (szeiger) [inline src/main/scala/scala/compat/java8/converterImpl/StepsArray.scala]: JVM arrays are covariant, so we can treat every object array as an `Array[AnyRef]` for the purpose of reading from it. Any "magical" Scala type (like `Unit` or `Null`) falls into that category, not matter if Scala treats it as a primitive or reference type. `Boolean` is the only primitive *JVM* type that gets mapped to `AnyStepper`, therefore it gets special treatment here. Since Scala, unlike the JVM, treats arrays as invariant, we do not have to cover the case where a primitive array could be seen as an object array.

**Comment by szeiger** [comment]
I ran a subset of the benchmark suite (fast summation on ArrayBuffer, Array and Vector). My original PR lost a few percent of performance in some cases but I couldn't pinpoint the reason. I saw bigger losses for the large test cases where the time spent constructing the Stepper should have less of an impact.

Measurements on the updated PR aren't fully conclusive, either. The Vector test with 10 elements is still a bit slower, but the one with ArrayBuffer is faster. I'd have to run more comprehensive tests on an idle machine but so far it looks like this doesn't have a negative impact on performance compared the old version (before this PR).

As the next step I want to try removing the specialized Steppers for boxed collections. I saw a few percent of performance degradation when I tried this with the old version. We could remove quite a bit of code by always going through AnyStepper in these cases but if the manually specialized versions turn out to still be measurably faster, they may be worth keeping.
